File: backend/src/services/effect_library_service.rs
// ✅ NOUVEAU: Service de gestion de la bibliothèque d'effets vidéo étendue

use crate::core::types::{AppError, AppResult};
use crate::models::effect_model::Effect;
use log::{info, warn};
use serde::{Deserialize, Serialize};
use sqlx::PgPool;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EffectSearchParams {
    pub category: Option<String>,
    pub tags: Option<Vec<String>>,
    pub search_query: Option<String>,
    pub is_premium: Option<bool>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EffectListResponse {
    pub effects: Vec<Effect>,
    pub total: i64,
    #[allow(dead_code)]
    pub limit: i64,
    #[allow(dead_code)]
    pub offset: i64,
}

pub struct EffectLibraryService {
    pool: PgPool,
}

impl EffectLibraryService {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    /// Récupère tous les effets avec filtres optionnels
    pub async fn list_effects(&self, params: EffectSearchParams) -> AppResult<EffectListResponse> {
        let limit = params.limit.unwrap_or(50).min(100); // Max 100
        let offset = params.offset.unwrap_or(0);

        let mut query_builder = sqlx::QueryBuilder::new(
            "SELECT id, name, category, description, ffmpeg_filter, parameters, tags, is_premium, popularity_score, created_at, updated_at FROM effects WHERE 1=1"
        );

        if let Some(category) = &params.category {
            query_builder.push(" AND category = ");
            query_builder.push_bind(category);
        }

        if let Some(ref tags) = params.tags {
            if !tags.is_empty() {
                query_builder.push(" AND tags && ");
                query_builder.push_bind(tags);
            }
        }

        if let Some(ref search_query) = params.search_query {
            if !search_query.is_empty() {
                query_builder.push(" AND (name ILIKE ");
                query_builder.push_bind(format!("%{}%", search_query));
                query_builder.push(" OR description ILIKE ");
                query_builder.push_bind(format!("%{}%", search_query));
                query_builder.push(")");
            }
        }

        if let Some(is_premium) = params.is_premium {
            query_builder.push(" AND is_premium = ");
            query_builder.push_bind(is_premium);
        }

        query_builder.push(" ORDER BY popularity_score DESC, name ASC");
        query_builder.push(" LIMIT ");
        query_builder.push_bind(limit);
        query_builder.push(" OFFSET ");
        query_builder.push_bind(offset);

        let effects: Vec<Effect> = query_builder
            .build_query_as()
            .fetch_all(&self.pool)
            .await
            .map_err(|e| {
                warn!("[EffectLibrary] Erreur requête effets: {}", e);
                AppError::Database(format!("Erreur récupération effets: {}", e))
            })?;

        // Compter le total
        let total_query = format!(
            "SELECT COUNT(*) FROM effects WHERE 1=1{}",
            build_where_clause(&params)
        );
        let total: i64 = sqlx::query_scalar(&total_query)
            .fetch_one(&self.pool)
            .await
            .unwrap_or(0);

        info!(
            "[EffectLibrary] ✅ {} effets trouvés (limit: {}, offset: {})",
            effects.len(),
            limit,
            offset
        );

        Ok(EffectListResponse {
            effects,
            total,
            limit,
            offset,
        })
    }

    /// Récupère un effet par son nom
    pub async fn get_effect_by_name(&self, name: &str) -> AppResult<Option<Effect>> {
        let effect = sqlx::query_as::<_, Effect>(
            "SELECT id, name, category, description, ffmpeg_filter, parameters, tags, is_premium, popularity_score, created_at, updated_at FROM effects WHERE name = $1"
        )
        .bind(name)
        .fetch_optional(&self.pool)
        .await
        .map_err(|e| {
            warn!("[EffectLibrary] Erreur récupération effet {}: {}", name, e);
            AppError::Database(format!("Erreur récupération effet: {}", e))
        })?;

        Ok(effect)
    }

    /// Récupère les effets par catégorie
    pub async fn get_effects_by_category(&self, category: &str) -> AppResult<Vec<Effect>> {
        let effects = sqlx::query_as::<_, Effect>(
            "SELECT id, name, category, description, ffmpeg_filter, parameters, tags, is_premium, popularity_score, created_at, updated_at FROM effects WHERE category = $1 ORDER BY popularity_score DESC, name ASC"
        )
        .bind(category)
        .fetch_all(&self.pool)
        .await
        .map_err(|e| {
            warn!("[EffectLibrary] Erreur récupération effets catégorie {}: {}", category, e);
            AppError::Database(format!("Erreur récupération effets: {}", e))
        })?;

        info!(
            "[EffectLibrary] ✅ {} effets trouvés pour catégorie {}",
            effects.len(),
            category
        );

        Ok(effects)
    }

    /// Recherche d'effets par tags
    pub async fn search_effects_by_tags(&self, tags: &[String]) -> AppResult<Vec<Effect>> {
        if tags.is_empty() {
            return Ok(vec![]);
        }

        let effects = sqlx::query_as::<_, Effect>(
            "SELECT id, name, category, description, ffmpeg_filter, parameters, tags, is_premium, popularity_score, created_at, updated_at FROM effects WHERE tags && $1 ORDER BY popularity_score DESC, name ASC"
        )
        .bind(tags)
        .fetch_all(&self.pool)
        .await
        .map_err(|e| {
            warn!("[EffectLibrary] Erreur recherche par tags: {}", e);
            AppError::Database(format!("Erreur recherche effets: {}", e))
        })?;

        info!(
            "[EffectLibrary] ✅ {} effets trouvés pour tags {:?}",
            effects.len(),
            tags
        );

        Ok(effects)
    }

    /// Met à jour le score de popularité d'un effet
    pub async fn update_popularity_score(&self, effect_id: i32, score: f64) -> AppResult<()> {
        sqlx::query("UPDATE effects SET popularity_score = $1 WHERE id = $2")
            .bind(score)
            .bind(effect_id)
            .execute(&self.pool)
            .await
            .map_err(|e| {
                warn!("[EffectLibrary] Erreur mise à jour popularité: {}", e);
                AppError::Database(format!("Erreur mise à jour: {}", e))
            })?;

        Ok(())
    }
}

fn build_where_clause(params: &EffectSearchParams) -> String {
    let mut clauses = Vec::new();

    if let Some(category) = &params.category {
        clauses.push(format!(" AND category = '{}'", category));
    }

    if let Some(ref tags) = params.tags {
        if !tags.is_empty() {
            let tags_str = tags
                .iter()
                .map(|t| format!("'{}'", t))
                .collect::<Vec<_>>()
                .join(",");
            clauses.push(format!(" AND tags && ARRAY[{}]", tags_str));
        }
    }

    if let Some(ref search_query) = params.search_query {
        if !search_query.is_empty() {
            clauses.push(format!(
                " AND (name ILIKE '%{}%' OR description ILIKE '%{}%')",
                search_query, search_query
            ));
        }
    }

    if let Some(is_premium) = params.is_premium {
        clauses.push(format!(" AND is_premium = {}", is_premium));
    }

    clauses.join("")
}

File: mobile/src/services/effectLibraryService.ts
// ✅ NOUVEAU: Service frontend pour bibliothèque d'effets vidéo (50+)

import { apiCall } from './api';

export interface Effect {
    id: number;
    name: string;
    category: 'transitions' | 'visual_effects' | 'animations' | 'special';
    description: string;
    ffmpeg_filter: string;
    parameters: Record<string, any>;
    tags: string[];
    is_premium: boolean;
    popularity_score: number;
    created_at: string;
    updated_at: string;
}

export interface EffectSearchParams {
    category?: string;
    tags?: string[];
    search_query?: string;
    is_premium?: boolean;
    limit?: number;
    offset?: number;
}

export interface EffectListResponse {
    effects: Effect[];
    total: number;
    limit: number;
    offset: number;
}

export interface EffectParameter {
    name: string;
    type: 'float' | 'int' | 'bool' | 'string' | 'color';
    min?: number;
    max?: number;
    default: any;
    description?: string;
}

export const effectLibraryService = {
    /**
     * Liste tous les effets avec filtres optionnels
     */
    async listEffects(params?: EffectSearchParams): Promise<EffectListResponse> {
        const queryParams = new URLSearchParams();

        if (params?.category) queryParams.append('category', params.category);
        if (params?.tags && params.tags.length > 0) queryParams.append('tags', params.tags.join(','));
        if (params?.search_query) queryParams.append('q', params.search_query);
        if (params?.is_premium !== undefined) queryParams.append('premium', String(params.is_premium));
        if (params?.limit) queryParams.append('limit', String(params.limit));
        if (params?.offset) queryParams.append('offset', String(params.offset));

        const queryString = queryParams.toString();
        const url = `/api/effects${queryString ? `?${queryString}` : ''}`;

        const response = await apiCall<EffectListResponse>(url);
        return response;
    },

    /**
     * Récupère un effet par son nom
     */
    async getEffectByName(name: string): Promise<Effect | null> {
        try {
            const response = await apiCall<Effect>(`/api/effects/${encodeURIComponent(name)}`);
            return response;
        } catch (error) {
            console.error(`[EffectLibrary] Erreur récupération effet ${name}:`, error);
            return null;
        }
    },

    /**
     * Récupère les effets par catégorie
     */
    async getEffectsByCategory(category: string): Promise<Effect[]> {
        try {
            const response = await apiCall<Effect[]>(`/api/effects/category/${encodeURIComponent(category)}`);
            return (response as any) || [];
        } catch (error) {
            console.error(`[EffectLibrary] Erreur récupération effets catégorie ${category}:`, error);
            return [];
        }
    },

    /**
     * Recherche d'effets par tags
     */
    async searchEffectsByTags(tags: string[]): Promise<Effect[]> {
        if (tags.length === 0) return [];

        try {
            const response = await effectLibraryService.listEffects({ tags });
            return response.effects || [];
        } catch (error) {
            console.error('[EffectLibrary] Erreur recherche par tags:', error);
            return [];
        }
    },

    /**
     * Recherche d'effets par query textuelle
     */
    async searchEffects(query: string, category?: string): Promise<Effect[]> {
        try {
            const response = await effectLibraryService.listEffects({
                search_query: query,
                category,
                limit: 50,
            });
            return response.effects || [];
        } catch (error) {
            console.error('[EffectLibrary] Erreur recherche:', error);
            return [];
        }
    },
};


